File: secretmanager/v1/python/google-cloud-secret-manager/secret-manager-client-secret-version-enable.py
# Copyright 2026 Google LLC
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.

# [START secretmanager_v1_secretmanagerservice_secretversion_enable]
import google.api_core.exceptions
from google.cloud import secretmanager_v1


def enable_secret_version(project_id: str, secret_id: str, version_id: str) -> None:
    """Enables a secret version. This demonstrates restoring access to a previously disabled secret version.

    Args:
        project_id: Google Cloud Project ID (such as 'example-project-id')
        secret_id: ID of the secret (such as 'my-secret-id')
        version_id: ID of the version (such as 'latest')
    """
    client = secretmanager_v1.SecretManagerServiceClient()

    name = f"projects/{project_id}/secrets/{secret_id}/versions/{version_id}"

    request = secretmanager_v1.EnableSecretVersionRequest(
        name=name,
    )

    try:
        response = client.enable_secret_version(request=request)
        print(f"Enabled secret version: {response.name}")
    except google.api_core.exceptions.NotFound:
        print(
            f"Error: The secret version '{name}' was not found. Check "
            "that the secret and version exist."
        )
    except Exception as e:
        print(f"An unexpected error occurred: {e}")


# [END secretmanager_v1_secretmanagerservice_secretversion_enable]

File: secretmanager/v1/nodejs/google-cloud-secret-manager/secret-manager-client-secret-version-enable.js
/**
 * Copyright 2026 Google LLC
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

'use strict';

// [START secretmanager_v1_secretmanagerservice_secretversion_enable]
// [START secretmanager_secretmanagerservice_secretversion_enable]
// [START secretmanager_enable_secret_version]

const {SecretManagerServiceClient} = require('@google-cloud/secret-manager');
const {status} = require('@grpc/grpc-js');

const client = new SecretManagerServiceClient();

/**
 * Enables a secret version, restoring access to a previously disabled version.
 * This demonstrates how to reactivate a secret version that was previously
 * disabled, making its payload accessible again.
 *
 * @param projectId Google Cloud Project ID (such as 'example-project-id')
 * @param secretId ID of the secret (such as 'my-secret-id')
 * @param versionId ID of the version (such as 'latest')
 */
async function enableSecretVersion(projectId, secretId, versionId) {
  const name = `projects/${projectId}/secrets/${secretId}/versions/${versionId}`;

  try {
    const [version] = await client.enableSecretVersion({
      name,
    });

    const dateObj = new Date(Number(version.createTime.seconds) * 1000);

    const createTime = new Intl.DateTimeFormat('en-US', {
      dateStyle: 'short',
      timeStyle: 'long',
      timeZone: 'GMT',
    }).format(dateObj);

    console.log(`Enabled secret version: ${version.name}`);
    console.log(`  Create Time: ${createTime}`);
    console.log(`  State: ${version.state}`);
    console.log(`  Replication: ${version.replicationStatus.replicationStatus}`);

  } catch (err) {
    if (err.code === status.FAILED_PRECONDITION) {
      console.error(
        `Error: The secret version '${name}' is already enabled. No action is required.`,
      );
    } else {
      console.error('An unexpected error occurred:', err);
    }
  }
}

// [END secretmanager_enable_secret_version]
// [END secretmanager_secretmanagerservice_secretversion_enable]
// [END secretmanager_v1_secretmanagerservice_secretversion_enable]

module.exports = {enableSecretVersion};
